migration.go

```diff
@@ -77,42 +85,62 @@ func (m *Migration) run(db *sql.DB, direction bool) error {
 		if !m.Registered {
 			return fmt.Errorf("ERROR %v: failed to run Go migration: Go functions must be registered and built into a custom binary (see https://github.com/pressly/goose/tree/master/examples/go-migrations)", m.Source)
 		}
-		tx, err := db.Begin()
-		if err != nil {
-			return fmt.Errorf("ERROR failed to begin transaction: %w", err)
-		}
-
 		fn := m.UpFn
+		fnNoTx := m.UpFnNoTx
 		if !direction {
 			fn = m.DownFn
+			fnNoTx = m.DownFnNoTx
+		}
+
+		if fnNoTx != nil {
+			// Run Go migration function with *sql.DB
+			if err := fnNoTx(db); err != nil {
+				return fmt.Errorf("ERROR %v: failed to run Go migration function %T: %w", filepath.Base(m.Source), fnNoTx, err)
+			}
+			if m.noVersioning {
+				return nil
+			}
+			if err := insertOrDeleteVersionNoTx(db, direction, m.Version); err != nil {
+				return fmt.Errorf("ERROR %v: failed to insert version with no tx: %w", filepath.Base(m.Source), err)
+			}
+			return nil
```

Comment:
you still returning early by the way. You'll never get to the 143 line, in which there are some useful logs

Reply:
That means I should delete 106 line "return nil", and then it will go to 143 line when everything is ok

Reply:
yeah

Reply:
and do the same a little below, at the end of the next if